datafusion/common/src/logical/eq.rs

```diff
@@ -0,0 +1,26 @@
+use arrow_schema::DataType;
+
+pub trait LogicallyEq<Rhs: ?Sized = Self> {
+    #[must_use]
+    fn logically_eq(&self, other: &Rhs) -> bool;
+}
+
+impl LogicallyEq for DataType {
+    fn logically_eq(&self, other: &Self) -> bool {
+        use DataType::*;
+
+        match (self, other) {
+            (Utf8 | LargeUtf8 | Utf8View, Utf8 | LargeUtf8 | Utf8View)
+            | (Binary | LargeBinary | BinaryView, Binary | LargeBinary | BinaryView) => {
+                true
+            }
+            (Dictionary(_, inner), other) | (other, Dictionary(_, inner)) => {
```

Comment:
How about both Dictionary cases?

Reply:
Technically it would work with the recursion but I could add a special case just to be sure

Reply:
We could avoid unnecessary call if we handle both Dict case